### PR TITLE
Make PodResources API include restartable init containers

### LINKED ISSUE #120501: PodResources API of kubelet does not report the resources allocated to restartable init containers
previous discussion: https://github.com/kubernetes/kubernetes/pull/120461#discussion_r1317515863

https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/pkg/kubelet/apis/podresources/server_v1.go#L51-L86

kubelet's PodResources API does not report the resources allocated to restartable init containers.
This should be fixed before we move to the beta stage.

/sig-node
/kind bug

/cc @ffromani @tzneal @SergeyKanzhelev

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Currently, the PodResources API of kubelet does not take restartable init containers into account.

This PR makes PodResource API include restartable init containers when the `SidecarContainers` feature gate is enabled.

#### Which issue(s) this PR fixes:

Fixes #120501 

#### Does this PR introduce a user-facing change?

```release-note
Node podresources API now includes init containers with containerRestartPolicy of `Always` when `SidecarContainers` feature is enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers
```

### CODE REVIEW

File: pkg/kubelet/apis/podresources/server_v1_test.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ import (
 	podresourcetest "k8s.io/kubernetes/pkg/kubelet/apis/podresources/testing"
 )
 
+var containerRestartPolicyAlways = v1.ContainerRestartPolicyAlways
```
Comment (ffromani) [inline pkg/kubelet/apis/podresources/server_v1_test.go:37]: can this be a variable local to a function?

**Review by ffromani — COMMENTED** [review]
thanks for the change. LGTM, but we should have some e2e test coverage. It will be probably sufficient to start with a e2e test running a pod with 1 init container, 1 sidecar container, 1 app container and check the endpoint returns the expected results.

**Comment by gjkim42** [comment]
> thanks for the change. LGTM, but we should have some e2e test coverage. It will be probably sufficient to start with a e2e test running a pod with 1 init container, 1 sidecar container, 1 app container and check the endpoint returns the expected results.


updated the existing e2e test but the related test is skipped due to the lack of cpu resources.
needs https://github.com/kubernetes/test-infra/pull/30844 to be merged.

**Review by ffromani — COMMENTED** [review]
/lgtm

File: pkg/kubelet/apis/podresources/server_v1.go [hunk 2]
```diff
@@ -23,8 +23,10 @@ import (
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	kubefeatures "k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/kubernetes/pkg/kubelet/types"
 
-	"k8s.io/kubelet/pkg/apis/podresources/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/kubelet/pkg/apis/podresources/v1"
```
Comment (ffromani) [inline pkg/kubelet/apis/podresources/server_v1.go:29]: not-so-nit: can we rename "podresourcesv1"? "v1" is usually (= almost always) an alias for "api/core/v1" so I'd keep the principle of least surprise

Comment (gjkim42) [reply pkg/kubelet/apis/podresources/server_v1.go]: done

**Review by ffromani — COMMENTED** [review]
LGTM
will have another look at the tests (ETA end of week February 16, worst case) for my peace of mind, but this should be ready now

**Comment by gjkim42** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120718/pull-kubernetes-node-e2e-containerd/1758083058650583040

I am not sure why pull-kubernetes-node-e2e-containerd failed.

**Review by ffromani — COMMENTED** [review]
/lgtm

thanks!

File: test/e2e_node/device_plugin_test.go [hunk 3]
```diff
@@ -687,21 +687,21 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 			gomega.Expect(resourcesForOurPod.Name).To(gomega.Equal(pod1.Name))
 			gomega.Expect(resourcesForOurPod.Namespace).To(gomega.Equal(pod1.Namespace))
 
-			// Note that the kubelet does not report resources for restartable
-			// init containers for now.
-			// See https://github.com/kubernetes/kubernetes/issues/120501.
-			// TODO: Fix this test to check the resources allocated to the
-			// restartable init container once the kubelet reports resources
-			// for restartable init containers.
```
Comment (gjkim42) [inline test/e2e_node/device_plugin_test.go:695]: https://github.com/kubernetes/kubernetes/pull/120461#discussion_r1317515863

removed this TODO and fixed the e2e test to check both restartable init container and regular container's device information as this PR is trying to address the #120501.
@ffromani

**Comment by gjkim42** [comment]
/assign @mrunalp 

Could you PTAL?

**Review by mrunalp — APPROVED** [review]

**Pull request merged**